### PR TITLE
Orphan blocks

### PR DESCRIPTION
**Pull request by almogdepaz**

### CODE REVIEW

File: mesh/block.go [hunk 1]
```diff
@@ -33,6 +34,7 @@ func NewExistingBlock(id BlockID, layerIndex LayerID, data []byte) *Block {
 	b := Block{
 		Id:         BlockID(id),
 		BlockVotes: make(map[BlockID]bool),
+		ViewEdges:  make(map[BlockID]bool),
```
Comment (antonlerner) [inline mesh/block.go]: we can drop using boolean and use struct{} instead, it's more efficient

File: mesh/encode.go [hunk 2]
```diff
@@ -2,22 +2,20 @@ package mesh
...
 
-func boolAsBytes(b bool) []byte {
-	var bitBool int8
-	if b {
-		bitBool = 1
-	}
-	return append(make([]byte, 0, 1), byte(bitBool))
+func uint32ToBytes(i uint32) []byte {
+	a := make([]byte, 4)
+	binary.LittleEndian.PutUint32(a, i)
```
Comment (antonlerner) [inline mesh/encode.go]: move to utility package

File: mesh/mesh.go [hunk 3]
```diff
@@ -48,17 +52,17 @@ func (m *mesh) IsContexuallyValid(b BlockID) bool {
 	return true
 }
 
-func (m *mesh) LatestIrreversible() uint32 {
-	return atomic.LoadUint32(&m.latestIrreversible)
+func (m *mesh) LocalLayer() uint32 {
```
Comment (antonlerner) [inline mesh/mesh.go:58]: what does local mean?

File: mesh/mesh.go [hunk 4]
```diff
@@ -81,45 +85,79 @@ func (m *mesh) AddLayer(layer *Layer) error {
...
-	m.SetLatestKnownLayer(uint32(layer.Index()))
+	atomic.AddUint32(&m.localLayer, 1)
+	m.SetLatestLayer(uint32(layer.Index()))
 	return nil
 }
 
 func (m *mesh) GetLayer(i LayerID) (*Layer, error) {
 	m.lMutex.RLock()
-	if i > LayerID(m.latestIrreversible) {
+	if i > LayerID(m.localLayer) {
```
Comment (antonlerner) [inline mesh/mesh.go:101]: I don't think there should be a limit on what getlayer should return, should we have an error if the later is not verified? if it's a helper method for this package make it private

File: mesh/mesh.go (hunk 3 above)
Comment (almogdepaz) [reply mesh/mesh.go:58]: tortoise is gonna add some counters it will be sorted as part of the next tortoise pr

File: mesh/mesh.go [hunk 5]
```diff
@@ -81,45 +85,79 @@ func (m *mesh) AddLayer(layer *Layer) error {
...
 	}
 	m.lMutex.RUnlock()
-	return m.mDB.getLayer(i)
+	return m.getLayer(i)
 }
 
 func (m *mesh) AddBlock(block *Block) error {
 	log.Debug("add block ", block.ID())
-	if err := m.mDB.addBlock(block); err != nil {
+	if err := m.addBlock(block); err != nil {
```
Comment (antonlerner) [inline mesh/mesh.go:112]: error or panic

File: mesh/mesh.go [hunk 6]
```diff
@@ -81,45 +85,79 @@ func (m *mesh) AddLayer(layer *Layer) error {
...
+	m.SetLatestLayer(uint32(block.Layer()))
+	//new block add to orphans
+	m.handleOrphanBlocks(block)
 	m.tortoise.HandleLateBlock(block) //todo should be thread safe?
 	return nil
 }
 
+//todo better thread safety
+func (m *mesh) handleOrphanBlocks(block *Block) {
+	m.orphanBlocks.Put(block.ID().ToBytes(), TRUE)
```
Comment (antonlerner) [inline mesh/mesh.go:125]: lock mutex?

File: mesh/mesh.go [hunk 7]
```diff
@@ -81,45 +85,79 @@ func (m *mesh) AddLayer(layer *Layer) error {
...
+	atomic.AddInt32(&m.orphanBlockCount, 1)
+	for b := range block.ViewEdges {
+		blockId := b.ToBytes()
+		if _, err := m.orphanBlocks.Get(blockId); err == nil {
```
Comment (antonlerner) [inline mesh/mesh.go:129]: what do we do when we get an error here?

File: mesh/mesh.go (hunk 4 above)
Comment (almogdepaz) [reply mesh/mesh.go:101]: good point, i think ill add a boolean and let the caller handle it

**Review by antonlerner — APPROVED** [review]

**Pull request merged**